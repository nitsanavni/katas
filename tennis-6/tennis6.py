class TennisGame6:
    def __init__(self, player1Name, player2Name):
        self.player1Name = player1Name
        self.player2Name = player2Name
        self.player1Score = 0
        self.player2Score = 0

    def won_point(self, playerName):
        if playerName == "player1":
            self.player1Score += 1
        else:
            self.player2Score += 1

    def score(self):
        result: str

        if self.player1Score == self.player2Score:
            # tie score
            tieScore: str
            match self.player1Score:
                case 0:
                    tieScore = "Love-All"
                case 1:
                    tieScore = "Fifteen-All"
                case 2:
                    tieScore = "Thirty-All"
                case _:
                    tieScore = "Deuce"

            result = tieScore
        elif self.player1Score >= 4 or self.player2Score >= 4:
            # end-game score
            endGameScore: str

            if self.player1Score - self.player2Score == 1:
                endGameScore = "Advantage " + self.player1Name
            elif self.player1Score - self.player2Score == -1:
                endGameScore = "Advantage " + self.player2Name
            elif self.player1Score - self.player2Score >= 2:
                endGameScore = "Win for " + self.player1Name
            else:
                endGameScore = "Win for " + self.player2Name

            result = endGameScore
        else:
            # regular score
            regularScore: str

            match (self.player1Score):
                case 0:
                    score1 = "Love"
                case 1:
                    score1 = "Fifteen"
                case 2:
                    score1 = "Thirty"
                case _:
                    score1 = "Forty"

            match (self.player2Score):
                case 0:
                    score2 = "Love"
                case 1:
                    score2 = "Fifteen"
                case 2:
                    score2 = "Thirty"
                case _:
                    score2 = "Forty"

            regularScore = score1 + "-" + score2

            result = regularScore

        return result
