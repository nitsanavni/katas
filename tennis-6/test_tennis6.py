from tennis6 import TennisGame6

from approvaltests import verify
from itertools import product


def test_tennis6():
    def score(p1, p2):
        game = TennisGame6("player1", "player2")
        for _ in range(p1):
            game.won_point("player1")
        for _ in range(p2):
            game.won_point("player2")
        return game.score()

    verify(
        "\n".join(
            [f"{p1}, {p2} -> {score(p1,p2)}" for p1, p2 in product(range(5), repeat=2)]
        )
    )
