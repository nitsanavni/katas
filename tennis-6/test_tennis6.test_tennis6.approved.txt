0, 0 -> Love-All
0, 1 -> Love-Fifteen
0, 2 -> Love-Thirty
0, 3 -> Love-Forty
0, 4 -> Win for player2
0, 5 -> Win for player2
1, 0 -> Fifteen-Love
1, 1 -> Fifteen-All
1, 2 -> Fifteen-Thirty
1, 3 -> Fifteen-Forty
1, 4 -> Win for player2
1, 5 -> Win for player2
2, 0 -> Thirty-Love
2, 1 -> Thirty-Fifteen
2, 2 -> Thirty-All
2, 3 -> Thirty-Forty
2, 4 -> Win for player2
2, 5 -> Win for player2
3, 0 -> Forty-Love
3, 1 -> Forty-Fifteen
3, 2 -> Forty-Thirty
3, 3 -> Deuce
3, 4 -> Advantage player2
3, 5 -> Win for player2
4, 0 -> Win for player1
4, 1 -> Win for player1
4, 2 -> Win for player1
4, 3 -> Advantage player1
4, 4 -> Deuce
4, 5 -> Advantage player2
5, 0 -> Win for player1
5, 1 -> Win for player1
5, 2 -> Win for player1
5, 3 -> Win for player1
5, 4 -> Advantage player1
5, 5 -> Deuce
